### DISCUSSION TITLE
Cross compile suing meson for arm Arch64

### DISCUSSION THREAD

**Discussion by semrade** [mesonbuild/meson#11292] [Q&A]
Hello
I want to cross-compile the Aravis librairy for the ARM architecture running on Linux.
The Aravis library use meson, you can find the repo link:https://github.com/AravisProject/aravis
I read the cross-compilation documentation to create the cross-compilation file,
```
[binaries]
c = '/opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-gcc'
cpp = '/opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-g++'
ar = '/opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-gcc-ar'
strip = '/opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-strip'
pkgconfig = '/usr/bin/aarch64-linux-gnu-pkg-config'
cmake='/usr/bin/cmake'

[properties]
#lib = '/usr/share/glib-2.0'
#c_stdlib = ['lib'] # Subproject name, variable name
c_stdlib = '/usr/share/glib-2.0'

[host_machine]
system = 'linux'
cpu_family = 'arm'
cpu = 'arm'
endian = 'little'
```

meson throws this error message but since I have never used the meson file am asking for help:
```
The Meson build system
Version: 0.63.3
Source dir: /home/labo-absen/Téléchargements/aravis-main
Build dir: /home/labo-absen/Téléchargements/aravis-main/build-mingw3
Build type: cross build
Project name: aravis
Project version: 0.8.23
C compiler for the host machine: /opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-gcc (gcc 8.2.1 "aarch64-linux-gnu-gcc (GNU Toolchain for the A-profile Architecture 8.2-2018-08 (arm-rel-8.23)) 8.2.1 20180802")
C linker for the host machine: /opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-gcc ld.bfd 8.2-2018
C++ compiler for the host machine: /opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-g++ (gcc 8.2.1 "aarch64-linux-gnu-g++ (GNU Toolchain for the A-profile Architecture 8.2-2018-08 (arm-rel-8.23)) 8.2.1 20180802")
C++ linker for the host machine: /opt/x86_64-aarch64-none-linux-gnu-8/gcc_arm_8_2_2018_08-x86_64_aarch64_linux_gnu/bin/aarch64-linux-gnu-g++ ld.bfd 8.2-2018
C compiler for the build machine: cc (gcc 11.3.0 "cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C linker for the build machine: cc ld.bfd 2.38
C++ compiler for the build machine: c++ (gcc 11.3.0 "c++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C++ linker for the build machine: c++ ld.bfd 2.38
Looking for a fallback subproject for the dependency c_stdlib because:
Use of fallback dependencies is forced.

meson.build:1:0: ERROR: Subproject name must not be an absolute path.

A full log can be found at /home/labo-absen/Téléchargements/aravis-main/build-mingw3/meson-logs/meson-log.txt

```

**Comment by dcbaker** [comment]
Your cpu_family should be aarch64, arm means 32bit arm, jfyi

**Reply by semrade** [reply]
Thank you I change it but still has this error 
Run-time dependency glib-2.0 found: NO (tried pkgconfig and cmake)

meson.build:59:0: ERROR: Dependency "glib-2.0" not found, tried pkgconfig and cmake

**Comment by eli-schwartz** [comment]
You should probably just skip the `c_stdlib` property, that's only really useful for example for embedded development, not normal cross compiling. You want to use the standard library provided with the cross compiler, which simply works out of the box.

**Reply by semrade** [reply]
Yes thank you for your help,
when I started the complication I didn't add the c_stdlib, I add it because I have an error so I thought maybe there is something missing:
meson.build:59:0: ERROR: Dependency "glib-2.0" not found, tried pkgconfig and cmake

**Reply by eli-schwartz** [reply]
Aha! That means that a third-party library dependency (not the C standard library) was not found. You'll need to either define `pkg_config_path` to the location where GLib (the gnome library, not the GNU standard C library) is installed, or run `meson wrap install glib` to import a vendored copy that will be built private to this project when a dependency isn't found.

**Reply by semrade** [reply]
OK, I prefer the `meson wrap install glib` version, do I run this command before the meson `meson setup --cross-file xxx.txt build-dir` is executed or after? I'm sorry I'm not familiar with meson and I don't really know about this compilation method.
could you explain more about the first point, please?  `third-party library dependency `
Thank you :)

**Reply by semrade** [reply]
after runing the meson wrip install glib this is the output 
`subprojects/libffi/src/meson.build:75:2: ERROR: Command "/opt/aarch64-non-linux-gnu/arm-gnu-toolchain-12.2.rel1-x86_64-aarch64-none-linux-gnu/bin/aarch64-none-linux-gnu-gcc -I/home/tarik/Downloads/aravis-0.8.22/build-mingw/subprojects/libffi/src/.. -I/home/tarik/Downloads/aravis-0.8.22/build-mingw/subprojects/libffi/src/../include -DARM -E -x assembler-with-cpp /home/tarik/Downloads/aravis-0.8.22/subprojects/libffi/src/../libffi.map.in -o /home/tarik/Downloads/aravis-0.8.22/build-mingw/subprojects/libffi/src/libffi.map" failed with status 1.`

the log is very long and I don't know if it is convenient to share her ...

**Reply by semrade** [reply]
```
A full log can be found at /home/tarik/Downloads/aravis-0.8.22/build-mingw/meson-logs/meson-log.txt
tarik@tarik-vm:~/Downloads/aravis-0.8.22$ cd build-mingw/
tarik@tarik-vm:~/Downloads/aravis-0.8.22/build-mingw$ meson compile -C build-mingw

ERROR: Current directory is not a meson build directory: `/home/tarik/Downloads/aravis-0.8.22/build-mingw/build-mingw`.
Please specify a valid build dir or change the working directory to it.
It is also possible that the build directory was generated with an old
meson version. Please regenerate it in this case.
tarik@tarik-vm:~/Downloads/aravis-0.8.22/build-mingw$ cd ..
tarik@tarik-vm:~/Downloads/aravis-0.8.22$ meson compile -C build-mingw

ERROR: Current directory is not a meson build directory: `/home/tarik/Downloads/aravis-0.8.22/build-mingw`.
Please specify a valid build dir or change the working directory to it.
It is also possible that the build directory was generated with an old
meson version. Please regenerate it in this case.
tarik@tarik-vm:~/Downloads/aravis-0.8.22$ 
```
